refactor: replace TaskMaster with Reactor

pollend opened the pull request:
### What this PR does

- replace one usage of `TaskMaster` with `GameScheduler`, creating and running `JPSImpl` on a separate thread
- removing the other usage of `TaskMaster` (via `TimeLimiter` logic)
  - _NOTE: We're not entirely clear yet about the performance impact of this, as path searches no longer time out after 3 seconds._

### Test Plan

Do the following once for each of the `FlexiblePathfinding#develop` and `FlexiblePathfinding#refactor/rework-concurrency-Reactivex` branches:

1. Start `CoreGameplay` world with added `WildAnimals` module and deps
2. Spawn 42 deer using the in-game console command `spawnPrefab deer`
3. Open in-game analytics (F3) and check FPS

### To Do

- [X]  test plan to check for potential performance impact in searching for valid paths

PR: (https://github.com/MovingBlocks/Terasology/issues/4798, https://github.com/MovingBlocks/Terasology/pull/4799)

pollend committed:
refactor: rework concurrency with reactiveX

PR: https://github.com/MovingBlocks/Terasology/issues/4798

DarkWeird commented on src/main/java/org/terasology/flexiblepathfinding/JPSImpl.java:
```diff
@@ -70,21 +69,17 @@ public static void setStatsEnabled(boolean statsEnabled) {
-            }
-        } else {
-            result = performSearch();
+        try {
```
1. rxjava have some error handling. It should be using there.
2. if you left try, then recordMetrics should be in finally block (have rx some thing there?)

jdrueckert replied:
went for (2), done.

DarkWeird commented on src/main/java/org/terasology/flexiblepathfinding/PathfinderSystem.java:
```diff
@@ -84,21 +78,42 @@ public int requestPath(EntityRef requester, Vector3i target, List<Vector3i> star
             entitiesWithPendingTasks.add(config.requester);
         }
+        Observable.just(new ConfigWrapper(callback, config))
+                .subscribeOn(GameThread.io())
```
Why `io()` ?

pollend replied:
its on a blocking thread where the computation is then passed off to another threading doing the actual computation.

jdrueckert replied:
`io()` is no longer used in the current state of the PR.

pollend committed:
rework with reactor

pollend committed:
update reactor

keturn committed:
chore: lint

keturn commented on src/main/java/org/terasology/flexiblepathfinding/PathfinderSystem.java:
```diff
@@ -84,20 +73,30 @@ public int requestPath(EntityRef requester, Vector3i target, List<Vector3i> star
+            }
+        } catch (InterruptedException e) {
+            // do nothing
+        }
```
This is scary, but I see it's just moved over from PathfinderTask.run, not introducing any new scariness.

keturn commented on src/main/java/org/terasology/flexiblepathfinding/PathfinderSystem.java:
```diff
@@ -84,20 +73,30 @@ public int requestPath(EntityRef requester, Vector3i target, List<Vector3i> star
-
-        PathfinderTask task = new PathfinderTask(world, config, callback);
-        taskMaster.offer(task);
+        Mono.just(new ConfigWrapper(callback, config))
+            .subscribeOn(GameScheduler.parallel())
+            .subscribe(this::processPath);
```
I think this would be equivalent:
```java
GameScheduler.parallel().schedule(() -> processPath(callback, config))
```
trading the ConfigWrapper tuple class for a lambda.

keturn commented on src/main/java/org/terasology/flexiblepathfinding/JPSImpl.java:
```diff
@@ -70,21 +63,17 @@ public static void setStatsEnabled(boolean statsEnabled) {
-        } else {
-            result = performSearch();
+        try {
+            boolean result = Boolean.TRUE.equals(Mono.fromCallable(this::performSearch)
+                    .subscribeOn(GameScheduler.parallel())
+                    .block(Duration.ofMillis((long) (config.maxTime * 1000.0f))));
```
We want to be very careful about when we call `block`, right?

This looks weird, because if I understand correctly, this `JPSImpl#run` is called by `PathfinderSystem#processPath`, which was _already_ sent to run on `GameScheduler.parallel()` by `PathfinderSystem#requestPath`.

So this is executing on the parallel scheduler, and it adds a new thing to that scheduler, and then _blocks one of that scheduler's threads_—that sounds like a good way to get a logjam, requiring later tasks to complete before earlier ones yield.

It also looks like the old implementation _only_ did that `timeLimiter` stuff **if** `config.executor` was non-null. Can we get away with dropping that feature?

pollend replied:
lets do that in another PR. this is just migrating the existing code over.

keturn replied:
Oh, I found where it set `config.executor`: it was in PathfinderSystem.requestPath. It looks like it did have a _second_ task queue for that purpose. That means this replacement where it's using the same scheduler for both is not equivalent.

jdrueckert replied:
I reverted the removal of the `TimeLimiter` usage as I don't understand why it is necessary for replacing `TaskMaster` with Reactor.
Also, I believe I'm using a new thread for the code path without the time limiter now, so the issue @skaldarnar reported below should be gone. At least when testing this in `CoreGameplay` + `WildAnimals` by spawning a deer, I didn't get any exceptions as far as I can see.

keturn reviewed (changes requested):
I think that once we resolve the concern I've noted with `JPSImpl.timeLimiter`, this is a fine stepping stone on the Reactor path. It'll let us move on with removing the `TaskMaster` class while not disturbing any more than it needs to.

If anyone is actually maintaining this module, they'll want to do another pass later that replaces the old callback-based interfaces with Reactor-compatible asynchronous functions.

pollend committed:
chore: update callback and config

keturn committed:
Merge remote-tracking branch 'origin/develop' into refactor/rework-concurrency-Reactivex

# Conflicts:
#	src/main/java/org/terasology/flexiblepathfinding/JPSImpl.java
#	src/main/java/org/terasology/flexiblepathfinding/PathfinderSystem.java
#	src/main/java/org/terasology/flexiblepathfinding/PathfinderTask.java
#	src/main/java/org/terasology/flexiblepathfinding/ShutdownTask.java

keturn commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Reactor: FlexiblePathfinding module](https://trello.com/c/2N1CGrlU/211-reactor-flexiblepathfinding-module)

skaldarnar commented:
Did a quick test run with after the migration with the `WildAnimal:deer` and `critter` behavior.

```
22:33:57.853 [main] DEBUG o.t.m.behaviors.actions.LogAction - Actor 228205: ---- critter ----
22:33:57.853 [main] DEBUG o.t.m.behaviors.actions.LogAction - Actor 228205: in stray Behavior
22:33:57.853 [main] DEBUG o.t.m.behaviors.actions.LogAction - Actor 228205: in doRandomMove Behavior
22:33:57.853 [main] DEBUG o.t.m.b.a.SetTargetToNearbyBlockNode - Actor 228205: in set_target_nearby_block Action
22:33:57.853 [main] DEBUG o.t.m.b.a.SetTargetToNearbyBlockNode - ... [228205]  start position: (-1.500E+1  3.000E+1 -1.000E+0)
22:33:57.853 [main] DEBUG o.t.m.b.a.SetTargetToNearbyBlockNode - ... [228205] target position: (-1.500E+1  3.000E+1 -1.000E+0) - distance: 0.0
22:33:57.853 [main] DEBUG o.t.m.behaviors.actions.LogAction - Actor 228205: in naiveMoveTo Behavior
22:33:57.853 [main] DEBUG o.t.m.b.actions.FindPathToNode - Actor 228205: construct find_path Action
22:33:57.853 [main] DEBUG o.t.m.b.actions.FindPathToNode - ... [228205]: compute path between (-1.500E+1  3.000E+1 -1.000E+0) -> (-1.500E+1  3.000E+1 -1.000E+0)
22:33:57.853 [main] DEBUG o.t.m.b.actions.FindPathToNode - Actor 228205: in find_path Action
22:33:57.853 [main] DEBUG o.t.m.b.actions.FindPathToNode - ... [228205]: ... still searching for path
22:33:57.853 [parallel-15] ERROR o.t.flexiblepathfinding.JPSImpl - Timeout of 10.0
java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread parallel-15
        at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:117)
        at reactor.core.publisher.Mono.block(Mono.java:1730)
        at org.terasology.flexiblepathfinding.JPSImpl.run(JPSImpl.java:69)
        at org.terasology.flexiblepathfinding.PathfinderSystem.processPath(PathfinderSystem.java:79)
        at org.terasology.flexiblepathfinding.PathfinderSystem.lambda$requestPath$0(PathfinderSystem.java:98)
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
        at io.micrometer.core.instrument.internal.TimedCallable.call(TimedCallable.java:46)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

The log about a "Timeout of 10.0" is misleading here (see correction in #26).

keturn commented:
That looks like what I was [worried about](https://github.com/Terasology/FlexiblePathfinding/pull/24#discussion_r743368053).

skaldarnar committed:
Merge remote-tracking branch 'origin/develop' into refactor/rework-concurrency-Reactivex

jdrueckert committed:
Merge branch 'develop' into refactor/rework-concurrency-Reactivex

jdrueckert committed:
chore: record metrics in finally block

jdrueckert committed:
revert: removal of TimeLimiter usage

keturn committed:
chore: lint 🚮

keturn commented:
Gave it a test run under JS.

The good news:
- it runs
- I see deer
- the deer move (**if** I dig them out from the ground they seem to be stuck in)
- the test suite passes

The bad news:
- it's good at using > 80% of _all_ CPUs when I'm just standing still and there are no chunks generating. When there _are_ chunks generating, it's unplayable.
- it ran out of memory and crashed in just under 15 minutes.

Point for comparison: In Core Gameplay, CPU avg across all CPUs is more like 10%, and memory usage is stable.

but I guess for PR purposes, the relevant point for comparison is the state of the same game configuration with and without the PR, not compared to Core Gameplay. :thinking:

keturn commented:
More testing:
- `develop` did not have the high-CPU-while-idle after the chunks were done generating
- then I tried loading the save from that game under this branch, and it seemed not-so-bad
- wondering if it was some kind of "new spawns vs loading from save" thing, I ran with `--create-last-game`
  - but wild animals spawns are not reproducible, so there weren't deer in sight, so I wasn't able to compare :weary:

jdrueckert commented:
@keturn Regarding spawns, you can always use `spawnPrefab deer` from the in-game console.

keturn commented on src/main/java/org/terasology/flexiblepathfinding/PathfinderSystem.java:
```diff
@@ -82,20 +72,30 @@ public int requestPath(EntityRef requester, Vector3i target, List<Vector3i> star
         }
-
-        if (config.executor == null) {
-            config.executor = workerTaskMaster.getExecutorService();
```
The question I have left on this review is: If this assignment of `config.executor` goes away, is there anything left using `config.executor` at all?

and if not, does that mean we never use TimeLimiter?

jdrueckert replied:
Probably that's the case (although we should double-check), yes.
Question for me is, if we don't use it, do we create a negative impact on performance? And if so, how can we model the timeout logic using Reactor / `GameScheduler`?

jdrueckert commented:
As discussed in today's contributor meeting: let's note down a test plan to check for potential performance impact in searching for valid paths and remove the time limiter logic with a comment to reintroduce it or remodel it with Reactor in case of issues.

jdrueckert committed:
chore: remove now unused timeout logic

- unclear performance impact without timeout logic
- if performance issues arise, reintroduce or reimplement with Reactor

jdrueckert committed:
chore: clean up imports

jdrueckert committed:
Merge remote-tracking branch 'origin/refactor/rework-concurrency-Reactivex' into refactor/rework-concurrency-Reactivex

jdrueckert commented:
So my first performance test plan was simply to spawn 42 deer in a CoreGameplay + WildAnimals module combo and look at the FPS...

Good news: With the current `FlexiblePathfinding` `develop`, I have 60 FPS
Bad news: With this branch checked out, I have 1.7 FPS - and no, the decimal point in there is no accident...

jdrueckert committed:
Merge branch 'develop' into refactor/rework-concurrency-Reactivex
